src/qw/docsection.py

```diff
@@ -0,0 +1,708 @@
+            at_end = predicate(node)
+            node.getparent().remove(node)
+            if at_end:
+                return next_node
+            node = next_node
+        return None
+
+    def _replace_field_runs(self, instr_node, plain_text: str):
+        """
+        Replace the fields containing instr_node with plain_text.
+
+        instr_node -- the <w:instrText> node within the field
+        plain_text -- replacement text
+        """
+        # Find the start of the field
+        r = instr_node.getparent()
+        self._delete_backwards_until(r.getprevious(), is_fld_begin)
+        # Find the normal text node
+        r = self._delete_forwards_until(r, is_fld_separate)
+        if r is not None:
+            # Remove everything from this run
+            for t in r.xpath("w:t"):
+                r.remove(t)
+            # Add our new text to this run
+            self._add_plain_text(r, plain_text)
+            # And delete the rest of the field furniture
+            self._delete_forwards_until(r.getnext(), is_fld_end)
+
+    def replace_field(self, name: str, plain_text: str) -> int:
+        """
+        Replace all MailMerge fields named `name` with pain text.
+
+        Returns the number of fields thus replaced.
+        """
+        count = 0
+        for instr in self.element[self.start_index].xpath("*/w:instrText"):
+            merge = MERGEFIELD_RE.fullmatch(instr.text)
+            if merge is not None and merge.group(1) == name:
+                self._replace_field_runs(instr, plain_text)
+                count += 1
+        return count
+
+
+class DocSectionParagraphReplacer(md.DocumentBuilder):
+    """
+    DocumentBuilder that replaces a single paragraph.
+
+    The paragraph in question is the first paragraph pointed to
+    by a DocSection object.
+    """
+
+    def __init__(self, section: DocSection):
+        """Initialize with a DocSection."""
+        self.section = section
+        self.first = True
+
+    def new_paragraph(
+        self,
+        paragraph_type: md.DocumentBuilder.ParagraphType = None,
+        paragraph_level: int = 0,
+    ):
+        """DocumentBuilder override."""
+        if self.first:
+            self.section.replace_first_paragraph(paragraph_type, paragraph_level)
+            self.first = False
+        else:
+            self.section.add_paragraph(paragraph_type, paragraph_level)
+
+    def add_run(
+        self,
+        text: str,
+        bold: md.DocumentBuilder.Boldness | None = None,
+        italic: md.DocumentBuilder.Italicness | None = None,
+        pre: md.DocumentBuilder.Spacing | None = None,
+    ):
+        """DocumentBuilder override."""
+        self.section.add_run(text, bold, italic, pre)
+
+    def add_hyperlink(
+        self,
+        text: str,
+        link: str,
+    ):
+        """DocumentBuilder override."""
+        self.section.add_hyperlink(text, link)
+
+    def end(self):
+        """DocumentBuilder override."""
```

Comment:
is this right, that end shouldn't do anything?

Reply:
Yeah, it's only needed for testing. It's a bit ugly.